### ISSUE TITLE
Allow list items to not split before a trailing line comment

### ISSUE THREAD

**Issue by munificent** [dart-lang/dart_style#139]
There are some really large collection literals that inevitably get split across lines. Often, though, splitting after *each* item is too gratuitous. Many of these literals in the original formatting have a few items on one line followed by a line comment, then the next line, etc.

It might be nice for the formatter to recognize this and allow the preceding items to not be split even though the list is in multiline mode. Something akin to treating a line comment as making the list's multisplit separable.

The tricky part is making sure this doesn't cancel out the "pre-emptive split" optimization we're doing for large collections.

Here's an example. Before:

```dart
  static const CATEGORIES = const <int>[
      OTHER, OTHER, OTHER, OTHER, OTHER, OTHER, OTHER, OTHER,       // 0-7
      OTHER, WHITESPACE, WHITESPACE, OTHER, OTHER, WHITESPACE,      // 8-13
      OTHER, OTHER, OTHER, OTHER, OTHER, OTHER, OTHER, OTHER,       // 14-21
      OTHER, OTHER, OTHER, OTHER, OTHER, OTHER, OTHER, OTHER,       // 22-29
      OTHER, OTHER, WHITESPACE,                                     // 30-32
      SYMBOL, OTHER, HASH, ALPHA, SYMBOL, SYMBOL, OTHER,            // !"#$%&´
      LPAREN, RPAREN, SYMBOL, SYMBOL, COMMA, SYMBOL, DOT, SYMBOL,   // ()*+,-./
      NUMERIC, NUMERIC, NUMERIC, NUMERIC, NUMERIC,                  // 01234
      NUMERIC, NUMERIC, NUMERIC, NUMERIC, NUMERIC,                  // 56789
      COLON, SEMICOLON, SYMBOL, SYMBOL, SYMBOL, QUERY, OTHER,       // :;<=>?@
      ALPHA, ALPHA, ALPHA, ALPHA, ALPHA, ALPHA, ALPHA, ALPHA,       // ABCDEFGH
      ALPHA, ALPHA, ALPHA, ALPHA, ALPHA, ALPHA, ALPHA, ALPHA,       // IJKLMNOP
      ALPHA, ALPHA, ALPHA, ALPHA, ALPHA, ALPHA, ALPHA, ALPHA,       // QRSTUVWX
      ALPHA, ALPHA, LSQUARE, OTHER, RSQUARE, SYMBOL, ALPHA, OTHER,  // YZ[\]^_'
      ALPHA, ALPHA, ALPHA, ALPHA, ALPHA, ALPHA, ALPHA, ALPHA,       // abcdefgh
      ALPHA, ALPHA, ALPHA, ALPHA, ALPHA, ALPHA, ALPHA, ALPHA,       // ijklmnop
      ALPHA, ALPHA, ALPHA, ALPHA, ALPHA, ALPHA, ALPHA, ALPHA,       // qrstuvwx
      ALPHA, ALPHA, LBRACE, SYMBOL, RBRACE, SYMBOL];                // yz{|}~
```

After:

```dart
  static const CATEGORIES = const <int>[
    OTHER,
    OTHER,
    OTHER,
    OTHER,
    OTHER,
    OTHER,
    OTHER,
    OTHER, // 0-7
    OTHER,
    WHITESPACE,
    WHITESPACE,
    OTHER,
    OTHER,
    WHITESPACE, // 8-13
    OTHER,
    OTHER,
    OTHER,
    OTHER,
    OTHER,
    OTHER,
    OTHER,
    OTHER, // 14-21
    OTHER,
    OTHER,
    OTHER,
    OTHER,
    OTHER,
    OTHER,
    OTHER,
    OTHER, // 22-29
    OTHER,
    OTHER,
    WHITESPACE, // 30-32
    SYMBOL,
    OTHER,
    HASH,
    ALPHA,
    SYMBOL,
    SYMBOL,
    OTHER, // !"#$%&´
    LPAREN,
    RPAREN,
    SYMBOL,
    SYMBOL,
    COMMA,
    SYMBOL,
    DOT,
    SYMBOL, // ()*+,-./
    NUMERIC,
    NUMERIC,
    NUMERIC,
    NUMERIC,
    NUMERIC, // 01234
    NUMERIC,
    NUMERIC,
    NUMERIC,
    NUMERIC,
    NUMERIC, // 56789
    COLON,
    SEMICOLON,
    SYMBOL,
    SYMBOL,
    SYMBOL,
    QUERY,
    OTHER, // :;<=>?@
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA, // ABCDEFGH
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA, // IJKLMNOP
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA, // QRSTUVWX
    ALPHA,
    ALPHA,
    LSQUARE,
    OTHER,
    RSQUARE,
    SYMBOL,
    ALPHA,
    OTHER, // YZ[\]^_'
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA, // abcdefgh
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA, // ijklmnop
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA,
    ALPHA, // qrstuvwx
    ALPHA,
    ALPHA,
    LBRACE,
    SYMBOL,
    RBRACE,
    SYMBOL
  ]; // yz{|}~
```

**Comment by seaneagan** [comment]
End of line comments generally apply to the entire line, so I think they are a a signal not to split the line.  Alternatively end of line comments could be moved to their own line before the line's comment which is now split, but that violates the whitespace only charter:

```static const CATEGORIES = const <int>[
    // 0-7
    OTHER,
    OTHER,
    OTHER,
    OTHER,
    OTHER,
    OTHER,
    OTHER,
    OTHER,
    // 8-13
    OTHER,
    WHITESPACE,
    WHITESPACE,
    OTHER,
    OTHER,
    WHITESPACE,
    // 14-21
    OTHER,
    OTHER
    // ...
  ];
```

**Comment by dynaxis** [comment]
Maybe it would be better to have a way to prevent dart_style from reformatting a code section. I encountered the exactly same issue with @munificent for a List literal with 256 entries. But in my case, it doesn't make too much sense to insert comments just to let the formatting intact.

Special comments would serve for that purpose quite well. The following is what works for Eclipse and Intellij IDEA.
```
// @formatter:off
...
// @formatter:on 
```

**Comment by munificent** [comment]
I really don't want to make it easy for people to "run" the formatter while effectively disabling it. We're encouraging teams to use the formatter for things like presubmit scripts so that they can ensure all code is formatted. That isn't effective if any team member can just add a `// @formatter:off` to the top of the file to silence the formatter.

In your case, @dynaxis, I think line comments would work OK. It could be as simple as:

```dart
var list = [
  0, 1, 2, 3, 4, 5, 6, 7, //
  0, 1, 2, 3, 4, 5, 6, 7, //
  0, 1, 2, 3, 4, 5, 6, 7, //
  0, 1, 2, 3, 4, 5, 6, 7, //
  0, 1, 2, 3, 4, 5, 6, 7, //
  0, 1, 2, 3, 4, 5, 6, 7, //
  0, 1, 2, 3, 4, 5, 6, 7, //
  0, 1, 2, 3, 4, 5, 6, 7, //
];
```

**Comment by dynaxis** [comment]
@munificent, I agree. It should be enough for me, too.

**Comment by munificent** [comment]
Oops, this wasn't the right bug to close.

**Comment by sebgeelen** [comment]
Is this fix present in Dart 1.11 ? 
If not what's the release version aimed ?

**Comment by munificent** [comment]
> Is this fix present in Dart 1.11 ? 

No, sorry, it didn't make it into that. dart_style 0.2.0 will be the next major release and it has a lot of changes. Because of that, I'm trying to shake out as many bugs as I can before I foist it onto the world, and there were a few left when 1.11 was cut.

I'm aiming to get it in 1.12.

**Comment by sebgeelen** [comment]
@munificent Thanks for the quick answer. Keep on the good work.
